sfa_dash/blueprints/form.py

```diff
@@ -1,42 +1,277 @@
+            self.api_handle = forecasts
+            self.formatter = self.forecast_formatter
+            self.metadata_template = 'data/metadata/site_metadata.html'
+        if data_type == 'observation':
+            self.template = 'forms/obs_form.html'
+            self.id_key = 'obs_id'
+            self.api_handle = observations
+            self.formatter = self.observation_formatter
+            self.metadata_template = 'data/metadata/site_metadata.html'
+        if data_type == 'site':
+            self.template = 'forms/site_form.html'
+            self.id_key = 'site_id'
+            self.api_handle = sites
+            self.formatter = self.site_formatter
 
-    def get(self):
-        json_dict = self.request_schema()
-        form_fields = {}
-        for key, value in json_dict[0]['definitions']['SiteSchema']['properties'].items():
-            form_fields[key] = {
-                'type': self.type_map.get(value['type'], None),
-                'label': value.get('title', key),
-                'name': key,
-                'description': value.get('description', None ),
-                'required': key in json_dict[0]['definitions']['SiteSchema']['required'],
-            }
-        form = render_template('forms/create_form.html', form_fields=form_fields)
-        return form
+    def flatten_dict(self, to_flatten):
+        """Flattens nested dictionaries, removing keys of the nested elements.
+        Useful for flattening API responses for prefilling forms on the
+        dashboard.
+        """
+        flattened = {}
+        for key, value in to_flatten.items():
+            if isinstance(value, dict):
+                flattened.update(self.flatten_dict(value))
+            else:
+                flattened[key] = value
+        return flattened
+
+    def parse_hhmm_field(self, data_dict, key_root):
+        """ Extracts and parses the hours and minutes inputs to create a
+        parseable time of day string in HH:MM format. These times are
+        displayed as two select fields designated with a name (key_root)
+        and _hours or _minutes suffix.
+
+        Parameters
+        ----------
+        data_dict: dict
+            Dictionary of posted form data
+
+        key_root: string
+            The shared part of the name attribute of the inputs to parse.
+            e.g. 'issue_time' will parse and concatenate 'issue_time_hours'
+            and 'issue_time_minutes'
+
+        Returns
+        -------
+        string
+            The time value in HH:MM format.
+        """
+        hours = int(data_dict[f'{key_root}_hours'])
+        minutes = int(data_dict[f'{key_root}_minutes'])
+        return f'{hours:02d}:{minutes:02d}'
+
+    def parse_timedelta(self, data_dict, key_root):
+        value = int(data_dict[f'{key_root}_number'])
+        units = data_dict[f'{key_root}_units']
+        if units == 'minutes':
+            return value
+        if units == 'hours':
+            return value * 60
+        else:
```

Comment:
are we allowing seconds? 

In any case, perhaps the `else` block should raise a `ValueError('invalid units')` instead of defaulting to something.

Reply:
I can allow seconds here if need be. I will have to allow for floats in the API as well.

Reply:
wait, what is this else case supposed to handle?

Reply:
It handles days. I'm going to update this to return 'NaT' as default so that the API returns an error with any others found in the form.

Reply:
I'm totally lost then. Maybe a doc string would help. Maybe it's supposed to be 1440 instead of 3600? Also use `if... elif... elif... else: NaT` to put it all into a single block.

Reply:
You're right. I updated this and the API to use minutes over seconds and messed up the values in both places. Thanks for catching it. 
Do you think there is a case where we need to allow for defining timedeltas in seconds, and therefore allowing the API to store them as floats with units of minutes?

Reply:
no, I don't think we need to support timedeltas of seconds. my comment about seconds was only because 3600 is the number of seconds in an hour, so I thought you might be handling seconds.